### ISSUE TITLE
Exceptions from Hosted UI (`UrlLauncherException`, `UserCancelledException`) cause debugger to pause

### ISSUE THREAD

**Issue by dkliss** [aws-amplify/amplify-flutter#4436]
### Description
More issues added below on exceptions:

```
Package: 
      amplify_flutter: ^1.6.1
      amplify_auth_cognito: ^1.6.2
```
ISSUE 1: Android

In my Android Emulator, there is no Chrome browzer installed.  As a result of this, when I initiate social sign in with any provider, the program is stuck at below exception and the exception does not reaches my try catch block. 
I expect the exception to be received in my try catch block so I can handle this.  

         String? recoverySuggestion;
          if (code == 'NOBROWSER') {
            recoverySuggestion =
                "Ensure you've added the <queries> tag to your "
                'AndroidManifest.xml as outlined in the docs';
          }
          throw UrlLauncherException(
            message ?? 'An unknown error occurred',
            recoverySuggestion: recoverySuggestion,
          );

I simply send out below within a try catch block. I was wondering if there is any reason I am not seeing above exception thrown 

```
try {
  final SignInResult signInResult = await Amplify.Auth.signInWithWebUI(
        provider: _getAuthProvider(signInOption),
      );
} catch (e) {
// I expect exception UrlLauncherException here. 
}
```
Expected: Exception should reach to app try catch block. 

ISSUE 2: iOS,

Below exception occurs but not sent to try catch bloc as well..

```
  throw const UserCancelledException(
              'The user cancelled the sign-in flow',
            );
```
Expected: Exception should reach to app try catch block. 

ISSUE 2: MacOS

When a browzer is closed, below exception is not received first time when web is closed but if web brozer is opened second time, i see below exception. 

    throw const UserCancelledException(
                'The user cancelled the sign-in flow',
              );

Expected: It should always occur when browzer is closed.   

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Steps to Reproduce

1. Send below
  
```
try {
final SignInResult signInResult = await Amplify.Auth.signInWithWebUI(
        provider: _getAuthProvider(signInOption),
      );
} catch (e) {}
```

2. Run code on Android where no browzer is installed.
3. See if UrlLauncherException exception is available in catch bloc.

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.16.9 

### Amplify Flutter Version

1.6.1

### Deployment Method

Amplify CLI

**Comment by Jordan-Nelson** [comment]
Hello - thanks for opening the issue. We will see if we can reproduce this with the info provided.

**Comment by Jordan-Nelson** [comment]
Hello - I attempted to reproduce the issue you are experiencing on Android and iOS and cannot reproduce it. I do see the break point hit. I have a few follow up questions:

- What IDE are you using?
- How are you launching/running the flutter app
- Can you confirm you are running in debug mode and have break points enabled?

**Comment by Jordan-Nelson** [comment]
@dkliss - Thanks for the detailed info. I think I had misunderstood the issue originally.

When I run an app in debug mode with "Uncaught Exceptions" enabled in VS code and follow the repro steps I do see the debugger stop in `/Users//hosted_ui_platform_flutter.dart` as you have noted. I can reproduce this with flutter version 3.16, but not flutter version 3.13. I think this is related to https://github.com/aws-amplify/amplify-flutter/issues/4345 and https://github.com/flutter/flutter/issues/141882 but I need to investigate a bit more.

**Comment by Jordan-Nelson** [comment]
I am going to label this as a bug for now, although at the moment I am unsure if this is a bug in Amplify Flutter or the Dart SDK.

**Comment by Jordan-Nelson** [comment]
@dkliss - I am fairly certain this is caused by https://github.com/flutter/flutter/issues/141882, which has been resolved in flutter master. I can reproduce this with Flutter v3.16, but cannot reproduce this with flutter master. Note, I also cannot reproduce this with flutter 3.13 (which is consistent with https://github.com/flutter/flutter/issues/141882).

I will leave this open until the Dart SDK fix gets cherry picked to a stable version of dart in case others are running into this issue. For now if you switch to flutter master you should not see the debugger pause in `/hosted_ui_platform_flutter.dart`.

**Comment by Jordan-Nelson** [comment]
The social provider page is not pushed on top of your app. The browser redirects to the social providers url. That is, from `localhost:3000` to `myaccounts.google.com` (for Google sign in). At that point your app is no longer active. Once authentication is complete the browser will redirect back to your application. It is correct that any code after `await Amplify.Auth.signInWithWebUI()` will not execute in this case.

> after I sign In and the redirect comes back to my app, my app has not received a token

I could not reproduce this. I am redirected back to `http://localhost:3000/?code=<code>&state=<state>` and have an active authentication session.

Are you seeing any errors? How do you know that there is not a successful sign in?

**Comment by Jordan-Nelson** [comment]
@dkliss this should be resolved in Dart version 3.4 which is included in Flutter v 3.20 and up. You can access Flutter 3.20 by using Flutter beta. This will likely land in Flutter stable in a couple months, but it depends on the Flutter teams release.

Since this is resolved in beta and on it's way to stable, I am going to close this issue out. If you face this issue with Flutter 3.20 or up please let me know and we can re-open the issue.
